framework/src/service/ServiceListeners.cpp

```diff
@@ -82,82 +93,171 @@ void ServiceListeners::Clear()
+{
+  // The following condition is true only if the listener is a non-static member function. In that case, if we already have a listener
+  // with the same ServiceListener and data, we remove that. Effectively, this replaces the existing listener with the new listener if
+  // it has the same ServiceListener and data.
+  //
+  // Note: With respect to the phrase "same ServiceListener", the comparison between two ServiceListener objects mostly returns true.
+  // This is because std::function objects aren't equality comparable. Hence, the only differentiating part is "data", which is the
+  // address of the object which has the non-static member function as the listener.
+  if (data != nullptr)
+  {
+    RemoveServiceListener(context, listener, data);
+  }
+
+  auto token = MakeListenerToken();
+  ServiceListenerEntry sle(context, listener, data, token.getId(), filter);
   {
     auto l = this->Lock(); US_UNUSED(l);
     serviceSet.insert(sle);
     CheckSimple_unlocked(sle);
   }
   coreCtx->serviceHooks.HandleServiceListenerReg(sle);
+  return token;
 }
 
-void ServiceListeners::RemoveServiceListener(const std::shared_ptr<BundleContextPrivate>& context, const ServiceListener& listener,
-                                             void* data)
+void ServiceListeners::RemoveServiceListenerEntry(ServiceListenerEntries::iterator it,
+                                                  ServiceListenerEntry& sle)
 {
-  ServiceListenerEntry entryToRemove(context, listener, data);
-  RemoveServiceListener(entryToRemove);
+  if (it != serviceSet.end())
+  {
+    sle = *it;
+    it->SetRemoved(true);
+    RemoveFromCache_unlocked(*it);
+    serviceSet.erase(it);
+  }
 }
 
-void ServiceListeners::RemoveServiceListener(const ServiceListenerEntry& entryToRemove)
+/**
+ * Called by the deprecated RemoveServiceListener(name_of_callable)
+ */
+void ServiceListeners::RemoveServiceListener(const std::shared_ptr<BundleContextPrivate>& context, const ServiceListener& listener,
+                                             void* data)
 {
   ServiceListenerEntry sle;
   {
     auto l = this->Lock(); US_UNUSED(l);
-    auto it = serviceSet.find(entryToRemove);
-    if (it != serviceSet.end())
-    {
-      sle = *it;
-      it->SetRemoved(true);
-      RemoveFromCache_unlocked(*it);
-      serviceSet.erase(it);
-    }
+    auto it = std::find_if(serviceSet.begin(), serviceSet.end(),
+                           [&context, &listener, &data](const ServiceListenerEntry& entry) {
+                             return entry.contains(context, data, listener);
+                           });
+    RemoveServiceListenerEntry(it, sle);
   }
   if (!sle.IsNull())
   {
     coreCtx->serviceHooks.HandleServiceListenerUnreg(sle);
   }
 }
 
-void ServiceListeners::AddBundleListener(const std::shared_ptr<BundleContextPrivate>& context, const BundleListener& listener, void* data)
+void ServiceListeners::RemoveServiceListener(const std::shared_ptr<BundleContextPrivate>& context,
```

Comment:
The only difference between the two ```RemoveServiceListener``` functions are the token ids. Is there a way to remove the code duplication?